### ISSUE TITLE
yubioath-desktop 5.0.4: error finding yubikey

### ISSUE THREAD

**Issue by djmaze** [NixOS/nixpkgs#104940]
**Describe the bug**
Since my last 20.09 update, yubioath-desktop does not recognize my yubikey (plugged via USB) anymore but shows error messages in the terminal. 

Start of the output:

```
Got library name:  "/nix/store/pd3xvxv3b9j6q704hp7h0lyr6i8wr8y4-pyotherside-1.5.9/lib/qt-5.15.0/qml/io/thp/pyotherside/libpyothersideplugin.so"
Qt Quick Layouts: Detected recursive rearrange. Aborting after two iterations.
Qt Quick Layouts: Detected recursive rearrange. Aborting after two iterations.
Qt Quick Layouts: Detected recursive rearrange. Aborting after two iterations.
"PyOtherSide error: Traceback (most recent call last):\n\n  File \"qrc:///py/yubikey.py\", line 12, in <module>\n    from ykman.descriptor import (\n\n  File \"/ni
x/store/d4wk69sgrv84bi1gfvar8ndpwm41dpmq-python3.8-yubikey-manager-3.1.1/lib/python3.8/site-packages/ykman/descriptor.py\", line 30, in <module>\n    from .util im
port PID, TRANSPORT, Mode, YUBIKEY\n\n  File \"/nix/store/d4wk69sgrv84bi1gfvar8ndpwm41dpmq-python3.8-yubikey-manager-3.1.1/lib/python3.8/site-packages/ykman/util.p
y\", line 41, in <module>\n    from OpenSSL import crypto\n\n  File \"/nix/store/kl2vfvl9abyh9i1vav7pl7jjqb33952y-python3.8-pyOpenSSL-19.1.0/lib/python3.8/site-pac
kages/OpenSSL/__init__.py\", line 8, in <module>\n    from OpenSSL import crypto, SSL\n\n  File \"/nix/store/kl2vfvl9abyh9i1vav7pl7jjqb33952y-python3.8-pyOpenSSL-1
9.1.0/lib/python3.8/site-packages/OpenSSL/crypto.py\", line 15, in <module>\n    from OpenSSL._util import (\n\n  File \"/nix/store/kl2vfvl9abyh9i1vav7pl7jjqb33952
y-python3.8-pyOpenSSL-19.1.0/lib/python3.8/site-packages/OpenSSL/_util.py\", line 6, in <module>\n    from cryptography.hazmat.bindings.openssl.binding import Bind
ing\n\n  File \"/nix/store/mgwlnh7mgx7fidav3xw0md8zj6b7rbzn-python3.8-cryptography-3.2.1/lib/python3.8/site-packages/cryptography/hazmat/bindings/openssl/binding.p
y\", line 220, in <module>\n    Binding.init_static_locks()\n\n  File \"/nix/store/mgwlnh7mgx7fidav3xw0md8zj6b7rbzn-python3.8-cryptography-3.2.1/lib/python3.8/site
-packages/cryptography/hazmat/bindings/openssl/binding.py\", line 155, in init_static_locks\n    __import__(\"_ssl\")\n\nModuleNotFoundError: No module named '_ssl
'\n"
Unhandled PyOtherSide error: Cannot import module: yubikey (Traceback (most recent call last):

  File "qrc:///py/yubikey.py", line 12, in <module>
    from ykman.descriptor import (

  File "/nix/store/d4wk69sgrv84bi1gfvar8ndpwm41dpmq-python3.8-yubikey-manager-3.1.1/lib/python3.8/site-packages/ykman/descriptor.py", line 30, in <module>
    from .util import PID, TRANSPORT, Mode, YUBIKEY

  File "/nix/store/d4wk69sgrv84bi1gfvar8ndpwm41dpmq-python3.8-yubikey-manager-3.1.1/lib/python3.8/site-packages/ykman/util.py", line 41, in <module>
    from OpenSSL import crypto

  File "/nix/store/kl2vfvl9abyh9i1vav7pl7jjqb33952y-python3.8-pyOpenSSL-19.1.0/lib/python3.8/site-packages/OpenSSL/__init__.py", line 8, in <module>
    from OpenSSL import crypto, SSL

  File "/nix/store/kl2vfvl9abyh9i1vav7pl7jjqb33952y-python3.8-pyOpenSSL-19.1.0/lib/python3.8/site-packages/OpenSSL/crypto.py", line 15, in <module>
    from OpenSSL._util import (

  File "/nix/store/kl2vfvl9abyh9i1vav7pl7jjqb33952y-python3.8-pyOpenSSL-19.1.0/lib/python3.8/site-packages/OpenSSL/_util.py", line 6, in <module>
    from cryptography.hazmat.bindings.openssl.binding import Binding

  File "/nix/store/mgwlnh7mgx7fidav3xw0md8zj6b7rbzn-python3.8-cryptography-3.2.1/lib/python3.8/site-packages/cryptography/hazmat/bindings/openssl/binding.py", line
 220, in <module>
    Binding.init_static_locks()

  File "/nix/store/mgwlnh7mgx7fidav3xw0md8zj6b7rbzn-python3.8-cryptography-3.2.1/lib/python3.8/site-packages/cryptography/hazmat/bindings/openssl/binding.py", line
 155, in init_static_locks
    __import__("_ssl")

ModuleNotFoundError: No module named '_ssl'
)
"PyOtherSide error: Traceback (most recent call last):\n\n  File \"<string>\", line 1, in <module>\n\nNameError: name 'yubikey' is not defined\n"
Unhandled PyOtherSide error: Function not found: 'yubikey.init' (Traceback (most recent call last):

  File "<string>", line 1, in <module>

NameError: name 'yubikey' is not defined
)
"PyOtherSide error: Traceback (most recent call last):\n\n  File \"<string>\", line 1, in <module>\n\nNameError: name 'yubikey' is not defined\n"
Unhandled PyOtherSide error: Function not found: 'yubikey.controller.check_descriptors' (Traceback (most recent call last):

  File "<string>", line 1, in <module>
```

**To Reproduce**
1. install yubioath-desktop from 20.09 stable (5.04, same problem with unstable as well) 
2. run `yubioath-desktop` in a terminal

**Expected behavior**

The command should not show any error output in the terminal and list my OATH keys in the GUI instead.

**Additional context**

I can successfully get information about the key using `ykman info`, so the yubikey itself seems to be working correctly. EDIT: `ykman oath list` works as well.

**Notify maintainers**
@mic92 @rissson

**Metadata**
```
 - system: `"x86_64-linux"`
 - host os: `Linux 5.4.78, NixOS, 20.09.2016.19db3e5ea27 (Nightingale)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.9`
 - channels(root): `"nixos-20.09.2016.19db3e5ea27"`
 - channels(martin): `"home-manager-20.09, nixos-20.09.2016.19db3e5ea27, nixpkgs-unstable-21.03pre253014.6625284c397"`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module:
```

**Comment by rissson** [comment]
I haven't had any problem with it (same version `5.0.4`, NixOS 20.09), with a Yubikey 5 NFC. What yubikey are you using?

**Comment by djmaze** [comment]
It's a NEO. But as I said, it worked before and it also works with `ykman`. 

Also, the error message `ModuleNotFoundError: No module named '_ssl'` sounds like a library / packaging problem to me?

**Comment by rissson** [comment]
On my machine, yubioath works as expected for both (19db3e5ea27 and 6625284c397) provided commits. Is there anything else that might be specific to your setup?

**Comment by djmaze** [comment]
Well, apart from using home-manager for user-specific packages (but not yubioath-desktop) and running i3 as window manager, I am not sure there is anything really special on my system. Also, I am mixing in a few app packages from unstable. 

I had some weird electron problems with i3 since the 20.09 upgrade (which have since been fixed upstream) and also getting notifications from notify-osd as dialog boxes instead of passive popups, but that's some minor annoyance I guess.

**Comment by Mic92** [comment]
Can you give us the full store path of your executable i.e.:

```console
$ which yubioath-desktop
/nix/store/1qd8rh2pln0kq0v7c64rs3zxk2x63naq-yubioath-desktop-5.0.4/bin/yubioath-desktop
```

Does the error happen in a pure nix-shell:

```console
$ nix-shell --pure -p yubioath-desktop
[nix-shell:~/git/nixpkgs-review]$ yubioath-desktop
```

**Comment by Mic92** [comment]
From the looks of it, it cannot `_ssl` from python's stdlib... which is weird.

**Comment by djmaze** [comment]
```bash
$ which yubioath-desktop
/home/martin/.nix-profile/bin/yubioath-desktop
$ stat -c %N /home/martin/.nix-profile/bin/yubioath-desktop
'/home/martin/.nix-profile/bin/yubioath-desktop' -> '/nix/store/glwqgxsx6cm7wsraia82b8n70vfqnjrn-yubioath-desktop-5.0.4/bin/yubioath-desktop'
```

@Mic92 When running a pure shell, it works! And there are no error messages anymore. Guess there is something wrong in my shell. For completeness, here is the output of the successful run:

```bash
[nix-shell:~]$ yubioath-desktop
QStandardPaths: XDG_RUNTIME_DIR not set, defaulting to '/run/user/1000/runtime-martin'
Got library name:  "/nix/store/rfa6i5camxmw8xyxm7kwdsgkvkq875vy-pyotherside-1.5.9/lib/qt-5.15.0/qml/io/thp/pyotherside/libpyothersideplugin.so"
...
```

**Comment by Mic92** [comment]
@djmaze now you can bisect your environment variables to figure out the breaking one. I suspect it has something to do with qt plugins that are loading from $HOME/.nix-profile and preventing python to load its own ssl library. You can export `LD_DEBUG=libs` to see if a different openssl is loaded before yours.

**Comment by djmaze** [comment]
Wow, I solved it, and that took quite some time.

First of, I got it running with a stripped down `PATH` consisting only of yubioath-desktop's bin path, leaving the rest of the environment unchanged:

```bash
env PATH=/nix/store/glwqgxsx6cm7wsraia82b8n70vfqnjrn-yubioath-desktop-5.0.4/bin yubioath-desktop
```
Using the `LD_DEBUG` approach running with the original `PATH` vs. the stripped-down one, I found out that the stripped down / pure version references a file called *_ssl.cpython-38-x86_64-linux-gnu.so* from the python3 package. That was somehow not in the original PATH.

Turns out I had *python3Minimal* package in my home-manager configuration. Making home-manager install the *python3* package instead, yubioath-desktop finally works as expected.

Thank you for the useful help and advice, much appreciated! :1st_place_medal:
